README.md
# fuser
Fused lasso for high-dimensional regression over groups. This package implements the model described in [Dondelinger et al. (2016)](https://arxiv.org/abs/1611.00953).

## Installation

```r
library('devtools')
install_github('FrankD/fuser')
```
## Example

See also the included vignette.

```r
library(fuser)
set.seed(123)

# Generate simple heterogeneous dataset
k = 4 # number of groups
p = 100 # number of covariates
n.group = 15 # number of samples per group
sigma = 0.05 # observation noise sd
groups = rep(1:k, each=n.group) # group indicators

# sparse linear coefficients
beta = matrix(0, p, k)
nonzero.ind = rbinom(p*k, 1, 0.025/k) # Independent coefficients
nonzero.shared = rbinom(p, 1, 0.025) # shared coefficients
beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25) 
beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)

X = lapply(1:k, function(k.i) matrix(rnorm(n.group*p),n.group, p)) # covariates 
y = sapply(1:k, function(k.i) X[[k.i]] %*% beta[,k.i] + rnorm(n.group, 0, sigma)) # response
X = do.call('rbind', X)

# Pairwise Fusion strength hyperparameters (tau(k,k'))
# Same for all pairs in this example
G = matrix(1, k, k) 

# Use L1 fusion to estimate betas (with near-optimal sparsity and 
# information sharing among groups)
beta.estimate = fusedLassoProximal(X, y, groups, lambda=0.001, tol=9e-5, 
                                   gamma=0.001, G, intercept=FALSE,
                                   num.it=2000) 

# Generate block diagonal matrices for L2 fusion approach
transformed.data = generateBlockDiagonalMatrices(X, y, groups, G)

# Use L2 fusion to estimate betas (with near-optimal information sharing among groups)
beta.estimate = fusedL2DescentGLMNet(transformed.data$X, transformed.data$X.fused, 
                                     transformed.data$Y, groups, lambda=c(0,0.001,0.1,1),
                                     gamma=0.001)
```



NEWS.md
# fuser 1.0.1

- Introduced option to scale sum-squared loss per subgroup by the subgroup sample size. This improves performance in situations with unbalanced groups.
- Fixed warning caused by deprecated function in the Matrix package.

# fuser 1.0.0

First full release of fuser package for subgroup regression with fusion 
penalties.

## Features

- L1 and L2 fusion penalized regression across heterogeneous subgroups.
- Ability to deal with high-dimensional datasets (tens of thousands of covariates).
- C implementation for additional speed gains.


tests/testthat.R
library(testthat)
library(fuser)

test_check("fuser")


tests/testthat/testRcpp.R
library(fuser)

# Test of Rcpp code for l1 fusion
set.seed(123)

# Generate simple heterogeneous dataset
k = 4 # number of groups
p = 10001 # number of covariates, high enough to trigger
          # calculation of cross-products on the fly
n.group = 500 # number of samples per group
sigma = 0.05 # observation noise sd
groups = rep(1:k, each=n.group) # group indicators

# sparse linear coefficients
beta = matrix(0, p, k)
nonzero.ind = rbinom(p*k, 1, 0.0025/k) # Independent coefficients
nonzero.shared = rbinom(p, 1, 0.0025) # shared coefficients
beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25) 
beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)

X = lapply(1:k, function(k.i) matrix(rnorm(n.group*p),n.group, p)) # covariates 
y = sapply(1:k, function(k.i) X[[k.i]] %*% beta[,k.i] + rnorm(n.group, 0, sigma)) # response
X = do.call('rbind', X)

# Pairwise Fusion strength hyperparameters (tau(k,k'))
# Same for all pairs in this example
G = matrix(1, k, k) 

# Use L1 fusion to estimate betas (with near-optimal sparsity and 
# information sharing among groups)
# Note, this will not converge in 500 iterations
beta.estimate = fusedLassoProximal(X, y, groups, lambda=1e-3, num.it=500,
                                   gamma=1e-2, G, intercept=FALSE) 

# Check that Rcpp code gives equivalent result
beta.estimate.2 = fusedLassoProximal(X, y, groups, lambda=1e-3, num.it=500,
                                   gamma=1e-2, G, intercept=FALSE, c.flag=TRUE) 

test_that("L1 Fusion: Compare Rcpp result with pure R code.", {
  expect_lt(mean((c(beta.estimate)-c(beta.estimate.2))^2), 0.005)
})



tests/testthat/testFusion.R
library(fuser)

# Test of using fusion approach to model heterogeneous
# datasets
set.seed(123)

# Generate simple heterogeneous dataset
k = 4 # number of groups
p = 100 # number of covariates
n.group = 15 # number of samples per group
sigma = 0.05 # observation noise sd
groups = rep(1:k, each=n.group) # group indicators

# sparse linear coefficients
beta = matrix(0, p, k)
nonzero.ind = rbinom(p*k, 1, 0.025/k) # Independent coefficients
nonzero.shared = rbinom(p, 1, 0.025) # shared coefficients
beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25) 
beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)

X = lapply(1:k, function(k.i) matrix(rnorm(n.group*p),n.group, p)) # covariates 
y = sapply(1:k, function(k.i) X[[k.i]] %*% beta[,k.i] + rnorm(n.group, 0, sigma)) # response
X = do.call('rbind', X)

# Pairwise Fusion strength hyperparameters (tau(k,k'))
# Same for all pairs in this example
G = matrix(1, k, k) 

# Use L1 fusion to estimate betas (with near-optimal sparsity and 
# information sharing among groups)
beta.estimate = fusedLassoProximal(X, y, groups, lambda=0.001, tol=9e-5, 
                                   gamma=0.001, G, intercept=FALSE,
                                   num.it=2000) 

test_that("L1 Fusion: Expect correlation between estimated and true beta.", {
  expect_gt(cor(c(beta.estimate), c(beta)), 0.9)
})

test_that("L1 Fusion: Expect small RMSE between estimated and true beta.", {
  expect_lt(mean((c(beta.estimate)-c(beta))^2), 0.01)
})

# Generate block diagonal matrices for L2 fusion approach
transformed.data = generateBlockDiagonalMatrices(X, y, groups, G)

# Use L2 fusion to estimate betas (with near-optimal information sharing among groups)
beta.estimate = fusedL2DescentGLMNet(transformed.data$X, transformed.data$X.fused, 
                                     transformed.data$Y, groups, lambda=c(0,0.001,0.1,1),
                                     gamma=0.001)


test_that("L2 Fusion: Expect correlation between estimated and true beta.", {
  expect_gt(cor(c(beta.estimate[,,2]), c(beta)), 0.9)
})

test_that("L2 Fusion: Expect small RMSE between estimated and true beta.", {
  expect_lt(mean((c(beta.estimate[,,2])-c(beta))^2), 0.005)
})




R/l2_fusion.R
# L2 Fusion approach optimization


#' Generate block diagonal matrices to allow for fused L2 optimization with glmnet.
#'
#' @param X covariates matrix (n by p).
#' @param Y response vector (length n).
#' @param groups vector of group indicators (ideally factors, length n)
#' @param G matrix representing the fusion strengths between pairs of
#' groups (K by K). Zero entries are assumed to be independent pairs.
#' @param intercept whether to include an (per-group) intercept in the model
#' @param penalty.factors vector of weights for the penalization of
#' each covariate (length p)
#' @param scaling Whether to scale each subgroup by its size. See Details for an explanation.
#'
#'
#' @return A list with components X, Y, X.fused and penalty, where
#' X is a n by pK block-diagonal bigmatrix, Y is a
#' re-arranged bigvector of length n, and X.fused is a
#' choose(K,2)*p by pK bigmatrix encoding the fusion penalties.
#'
#' @details We use the \code{glmnet} package to perform fused subgroup regression.
#' In order to achieve this, we need to reformulate the problem as Y' = X'beta',
#' where Y' is a concatenation of the responses Y and a vector of zeros, X' is a
#' a matrix consisting of the block-diagonal matrix n by pK matrix X, where each
#' block contains the covariates for one subgroups, and the choose(K,2)*p by pK
#' matrix encoding the fusion penalties between pairs of groups. The vector of
#' parameters beta' of length pK can be rearranged as a p by K matrix giving the
#' parameters for each subgroup. The lasso penalty on the parameters is handled
#' by glmnet.
#'
#' One weakness of the approach described above is that larger subgroups will
#' have a larger influence on the global parameters lambda and gamma.
#' In order to mitigate this, we introduce the \code{scaling} parameter. If
#' \code{scaling=TRUE}, then we scale the responses and covariates for each
#' subgroup by the number of samples in that group.
#'
#'
#' @import Matrix
#'
#' @export
#'
#' @examples
#' set.seed(123)
#'
#' # Generate simple heterogeneous dataset
#' k = 4 # number of groups
#' p = 100 # number of covariates
#' n.group = 15 # number of samples per group
#' sigma = 0.05 # observation noise sd
#' groups = rep(1:k, each=n.group) # group indicators
#' # sparse linear coefficients
#' beta = matrix(0, p, k)
#' nonzero.ind = rbinom(p*k, 1, 0.025/k) # Independent coefficients
#' nonzero.shared = rbinom(p, 1, 0.025) # shared coefficients
#' beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25)
#' beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)
#'
#' X = lapply(1:k,
#'            function(k.i) matrix(rnorm(n.group*p),
#'                                 n.group, p)) # covariates
#' y = sapply(1:k,
#'            function(k.i) X[[k.i]] %*% beta[,k.i] +
#'                            rnorm(n.group, 0, sigma)) # response
#' X = do.call('rbind', X)
#'
#' # Pairwise Fusion strength hyperparameters (tau(k,k'))
#' # Same for all pairs in this example
#' G = matrix(1, k, k)
#'
#' # Generate block diagonal matrices
#' transformed.data = generateBlockDiagonalMatrices(X, y, groups, G)
generateBlockDiagonalMatrices <- function(X, Y, groups, G, intercept=FALSE,
                                          penalty.factors=rep(1, dim(X)[2]),
                                          scaling=TRUE) {
  group.names = sort(unique(groups))
  num.groups = length(group.names)

  num.pairs = num.groups*(num.groups-1)/2

  if(intercept) X = cbind(X, matrix(1, dim(X)[1], 1)) # Include intercept

  new.y = Matrix(0, length(Y)+num.pairs*dim(X)[2],1, sparse=TRUE)
  new.x = Matrix(0, dim(X)[1], dim(X)[2]*num.groups,
                 sparse=TRUE)
  new.x.f = Matrix(0, num.pairs*dim(X)[2], dim(X)[2]*num.groups,
                   sparse=TRUE)

  row.start = 1
  col.start = 1

  # Add data into block matrix
  for(group.i in 1:num.groups) {
    group.inds = groups==group.names[group.i]

    row.range = row.start:(row.start+sum(group.inds)-1)
    col.range = col.start:(col.start+dim(X)[2]-1)

    new.y[row.range] = Y[group.inds]
    new.x[row.range, col.range] = X[group.inds,]

    if(scaling) {
      new.y[row.range] = new.y[row.range]/sum(group.inds)
      new.x[row.range, col.range] =
        new.x[row.range, col.range]/sum(group.inds)
    }

    row.start = row.start + sum(group.inds)
    col.start = col.start + dim(X)[2]
  }

  col.start.i = 1
  row.start = 1

  # Add gamma contraints into block matrix
  for(group.i in 1:(num.groups-1)) {

    col.start.j = col.start.i + dim(X)[2]

    col.range.i = col.start.i:(col.start.i+dim(X)[2]-1)

    for(group.j in (group.i+1):num.groups) {

      tau = G[group.i, group.j]

      row.range = row.start:(row.start+dim(X)[2]-1)
      col.range.j = col.start.j:(col.start.j+dim(X)[2]-1)

      new.x.f[cbind(row.range, col.range.i)] = sqrt(tau)
      new.x.f[cbind(row.range, col.range.j)] = -sqrt(tau)

      if(intercept) { # Don't fuse intercept
        new.x.f[row.range[length(row.range)],
                col.range.i[length(col.range.i)]] = 0
        new.x.f[row.range[length(row.range)],
                col.range.j[length(col.range.j)]] = 0
      }

      row.start = row.start + dim(X)[2]
      col.start.j = col.start.j + dim(X)[2]
    }

    col.start.i = col.start.i + dim(X)[2]
  }

  if(intercept) {
    penalty = rep(c(penalty.factors, 0), num.groups)
  } else {
    penalty = rep(penalty.factors, num.groups)
  }

  return(list(X=new.x, Y=new.y, X.fused=new.x.f, penalty=penalty))
}

#' Optimise the fused L2 model with glmnet (using transformed input data)
#'
#' @param transformed.x Transformed covariates (output of generateBlockDiagonalMatrices)
#' @param transformed.x.f Transformed fusion constraints (output of generateBlockDiagonalMatrices)
#' @param transformed.y Transformed response (output of generateBlockDiagonalMatrices)
#' @param groups Grouping factors for samples (a vector of size n, with K factor levels)
#' @param lambda Sparsity penalty hyperparameter
#' @param gamma Fusion penalty hyperparameter
#' @param ... Further options passed to glmnet.
#'
#' @return Matrix of fitted beta values.
#' @import glmnet
#' @export
#' @return
#' A matrix with the linear coefficients for each group (p by k).
#'
#' @examples
#'
#' #' set.seed(123)
#'
#' # Generate simple heterogeneous dataset
#' k = 4 # number of groups
#' p = 100 # number of covariates
#' n.group = 15 # number of samples per group
#' sigma = 0.05 # observation noise sd
#' groups = rep(1:k, each=n.group) # group indicators
#' # sparse linear coefficients
#' beta = matrix(0, p, k)
#' nonzero.ind = rbinom(p*k, 1, 0.025/k) # Independent coefficients
#' nonzero.shared = rbinom(p, 1, 0.025) # shared coefficients
#' beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25)
#' beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)
#'
#' X = lapply(1:k,
#'            function(k.i) matrix(rnorm(n.group*p),
#'                                 n.group, p)) # covariates
#' y = sapply(1:k,
#'            function(k.i) X[[k.i]] %*% beta[,k.i] +
#'                            rnorm(n.group, 0, sigma)) # response
#' X = do.call('rbind', X)
#'
#' # Pairwise Fusion strength hyperparameters (tau(k,k'))
#' # Same for all pairs in this example
#' G = matrix(1, k, k)
#'
#' # Generate block diagonal matrices
#' transformed.data = generateBlockDiagonalMatrices(X, y, groups, G)
#'
#' # Use L2 fusion to estimate betas (with near-optimal information
#' # sharing among groups)
#' beta.estimate = fusedL2DescentGLMNet(transformed.data$X,
#'                                      transformed.data$X.fused,
#'                                      transformed.data$Y, groups,
#'                                      lambda=c(0,0.001,0.1,1),
#'                                      gamma=0.001)
fusedL2DescentGLMNet <- function(transformed.x, transformed.x.f,
                                 transformed.y, groups, lambda, gamma=1,
                                 ...) {

  # Incorporate fusion penalty global hyperparameter
  transformed.x.f = transformed.x.f * sqrt(gamma *
                             (dim(transformed.x)[1] + dim(transformed.x.f)[1]))
  transformed.x = rbind(transformed.x, transformed.x.f)

  group.names = sort(unique(groups))
  num.groups = length(group.names)

  glmnet.result = glmnet(transformed.x, transformed.y, standardize=FALSE, ...)

  beta.mat = array(NA, c(dim(transformed.x)[2]/num.groups, num.groups, length(lambda)))

  for(lambda.i in 1:length(lambda)) {

    coef.temp = coef.glmnet(glmnet.result,
                     s=lambda[lambda.i]*length(groups)/dim(transformed.x)[1]) # Correction for extra dimensions
    beta.mat[,,lambda.i] = coef.temp[2:length(coef.temp)]
  }

  return(beta.mat)
}


src/crossprod.cpp
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-

// we only include RcppEigen.h which pulls Rcpp.h in for us
#include <RcppEigen.h>

using namespace std;
using Eigen::Map;
using Eigen::MatrixXd;
using Eigen::Lower;
using Rcpp::as;
// via the depends attribute we tell Rcpp to create hooks for
// RcppEigen so that the build process will know what to do
//
// [[Rcpp::depends(RcppEigen)]]


// Calculate crossproduct XtX
//MatrixXd crossprodCPP(const MatrixXd& X) {
  

//}


// Calculate crossproducts XtX and XtY and return product with 
// Beta matrix
// [[Rcpp::export]]
Eigen::MatrixXd doubleCrossProd(Eigen::MatrixXd A, Eigen::MatrixXd B) {
  //cout << A << " ";
  //const Map<MatrixXi> A(as<Map<MatrixXi> >(AA);
  MatrixXd AtAB = A.adjoint() * (A * B);

  return AtAB;
}






src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppEigen.h>
#include <Rcpp.h>

using namespace Rcpp;

// getNumberNativeIterationsTaken
NumericVector getNumberNativeIterationsTaken();
RcppExport SEXP _fuser_getNumberNativeIterationsTaken() {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    rcpp_result_gen = Rcpp::wrap(getNumberNativeIterationsTaken());
    return rcpp_result_gen;
END_RCPP
}
// genFusedLassoProximal_loop
Eigen::MatrixXd genFusedLassoProximal_loop(Nullable<List> XX, Nullable<List> XY, Nullable<List> X_list, Eigen::VectorXd Y, NumericVector samp_sizes, Eigen::MatrixXd C, bool intercept, int p, int k, int num_iters, Nullable<NumericVector> penalty_factors, double L_U_inv, Eigen::MatrixXd B_old, double mu, Eigen::MatrixXd W, Eigen::MatrixXd weighted_delta_f, double tol);
RcppExport SEXP _fuser_genFusedLassoProximal_loop(SEXP XXSEXP, SEXP XYSEXP, SEXP X_listSEXP, SEXP YSEXP, SEXP samp_sizesSEXP, SEXP CSEXP, SEXP interceptSEXP, SEXP pSEXP, SEXP kSEXP, SEXP num_itersSEXP, SEXP penalty_factorsSEXP, SEXP L_U_invSEXP, SEXP B_oldSEXP, SEXP muSEXP, SEXP WSEXP, SEXP weighted_delta_fSEXP, SEXP tolSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< Nullable<List> >::type XX(XXSEXP);
    Rcpp::traits::input_parameter< Nullable<List> >::type XY(XYSEXP);
    Rcpp::traits::input_parameter< Nullable<List> >::type X_list(X_listSEXP);
    Rcpp::traits::input_parameter< Eigen::VectorXd >::type Y(YSEXP);
    Rcpp::traits::input_parameter< NumericVector >::type samp_sizes(samp_sizesSEXP);
    Rcpp::traits::input_parameter< Eigen::MatrixXd >::type C(CSEXP);
    Rcpp::traits::input_parameter< bool >::type intercept(interceptSEXP);
    Rcpp::traits::input_parameter< int >::type p(pSEXP);
    Rcpp::traits::input_parameter< int >::type k(kSEXP);
    Rcpp::traits::input_parameter< int >::type num_iters(num_itersSEXP);
    Rcpp::traits::input_parameter< Nullable<NumericVector> >::type penalty_factors(penalty_factorsSEXP);
    Rcpp::traits::input_parameter< double >::type L_U_inv(L_U_invSEXP);
    Rcpp::traits::input_parameter< Eigen::MatrixXd >::type B_old(B_oldSEXP);
    Rcpp::traits::input_parameter< double >::type mu(muSEXP);
    Rcpp::traits::input_parameter< Eigen::MatrixXd >::type W(WSEXP);
    Rcpp::traits::input_parameter< Eigen::MatrixXd >::type weighted_delta_f(weighted_delta_fSEXP);
    Rcpp::traits::input_parameter< double >::type tol(tolSEXP);
    rcpp_result_gen = Rcpp::wrap(genFusedLassoProximal_loop(XX, XY, X_list, Y, samp_sizes, C, intercept, p, k, num_iters, penalty_factors, L_U_inv, B_old, mu, W, weighted_delta_f, tol));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_fuser_getNumberNativeIterationsTaken", (DL_FUNC) &_fuser_getNumberNativeIterationsTaken, 0},
    {"_fuser_genFusedLassoProximal_loop", (DL_FUNC) &_fuser_genFusedLassoProximal_loop, 17},
    {NULL, NULL, 0}
};

RcppExport void R_init_fuser(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/l1_fusion.R
.pkg.env <- new.env(parent = emptyenv())
.pkg.env$lastNumIters <- NA

#' Big eigenvalue calculation
#'
#' Calculate maximal eigenvalue of \code{t(X) \%*\% X} for big
#' matrices using singular value decomposition.
#'
#' @param X matrix to be evaluated (can be a Matrix object).
#' @param method One of 'irlba' or 'RSpectra'
#'
#' @return The maximal eigenvalue.
#'
#' @export
#' @import irlba
#' @import RSpectra
bigeigen <- function(X, method = "RSpectra") {
  if (method == "irlba") {
    # Define multiplication function
    matmul <- function(A, B) {
      # Bigalgebra requires matrix/vector arguments
      if (is.null(dim(B)))
        B = cbind(B)
      return(cbind( (A %*% B)[]) )
    }

    max.eigen = (irlba(X, 1, 1, mult = matmul)$d[1]) ^ 2
  } else {
    max.eigen = (svds(X, 1)$d[1]) ^ 2
  }

  return(max.eigen)
}

#' Following a call to fusedLassoProximal, returns the actual number of iterations taken.
#'
#' @export
#'
#' @return Number of iterations performed in the previous call to fusedLassoProximal.
fusedLassoProximalIterationsTaken <- function() {
  .pkg.env$lastNumIters
}


#' Fused lasso optimisation with the proximal-gradient method (Chen et al. 2010).
#' Internal method
#'
#' @param X matrix of covariates (n by p)
#' @param Y vector of responses (length n)
#' @param groups vector of group indicators (length n)
#' @param group.names group names (length k)
#' @param XX \code{t(X) \%*\% X}
#' @param XY \code{t(X) \%*\% Y}
#' @param p Number of covariates
#' @param k Number of groups
#' @param samp.sizes Sample size for each group
#' @param lambda Sparsity hyperparameter
#' @param gamma Fusion hyperparameter
#' @param C constraints
#' @param G pairwise fusion hyperparameters (tau)
#' @param epsilon step size
#' @param tol tolerance
#' @param num.it number of iterations
#' @param lam.max maximal eigenvalue (pre-calculated)
#' @param intercept whether to include intercept or not
#' @param penalty.factors setting l1 penalty factors for different coefficients
#' @param X.list X as a list (required if lam.max calculated on the fly)
#' @param c.flag C++ subroutines will be used if true.
#'
#' @return
#'
#' @useDynLib fuser, .registration=TRUE
#' @importFrom Rcpp evalCpp
#'
.genFusedLassoProximal <- function(X, Y, groups, group.names = sort(unique(groups)),
                                         XX = NULL, XY, p, k, samp.sizes, lambda, gamma,
                                         C, G, epsilon = 1e-04 * p * dim(C)[2],
                                         tol = 1e-06, num.it = 1000, lam.max = NULL,
                                         intercept = TRUE, penalty.factors = NULL,
                                         X.list = NULL, c.flag = FALSE) {
  n.c = dim(C)[2]  # Number of constraints, including sparsity

  D = 0.5 * p * n.c

  mu = epsilon / (2 * D)  # smoothness parameter

  if (is.null(lam.max)) {
    if (!is.null(XX)) {
      lam.max = sum(sapply(XX, function(x)
        max(
          eigen(x, symmetric = TRUE, only.values = TRUE)$values
        )))  # maximal eigenvalue
    } else {
      lam.max = sum(sapply(1:length(group.names), function(g)
        bigeigen(X.list[[g]])))  # maximal eigenvalue
    }

    G.temp = G
    G.temp[lower.tri(G)] = 0

    L_U = lam.max + (lambda ^ 2 + 2 * gamma ^ 2 * max(colSums(G.temp))) /
      mu
  } else {
    G.temp = G
    G.temp[lower.tri(G)] = 0
    L_U = lam.max + (lambda ^ 2 + 2 * gamma ^ 2 * max(colSums(G.temp))) /
      mu
  }

  gc()

  L_U.inv = 1 / L_U


  if (intercept) {
    W = matrix(0, p + 1, k)
    B.old = matrix(0, p + 1, k)
    weighted.delta.f = matrix(0, p + 1, k)
  } else {
    W = matrix(0, p, k)
    B.old = matrix(0, p, k)
    weighted.delta.f = matrix(0, p, k)
  }


  if (c.flag) {
    # Call native function

    group.ordering = order(groups)
    Y.sorted = Y[group.ordering]

    result = genFusedLassoProximal_loop(
      XX, XY, X.list, Y.sorted, samp.sizes, C, intercept, p, k,
      num.it, penalty.factors, L_U.inv, B.old, mu, W, weighted.delta.f, tol
    )
    .pkg.env$lastNumIters = getNumberNativeIterationsTaken()

    colnames(result) = group.names

    return(result)
  } else {
    # Call R function
    return(.genFusedLassoProximal.R(Y, groups, group.names, XX, XY, p, k, samp.sizes, lambda, gamma,
                                    C, G, epsilon, tol, num.it, intercept, penalty.factors, X.list,
                                    L_U.inv, B.old, mu, W, weighted.delta.f))
  }
}






.genFusedLassoProximal.R <- function(Y, groups, group.names, XX, XY, p, k, samp.sizes, lambda, gamma,
                                     C, G, epsilon, tol, num.it, intercept, penalty.factors, X.list,
                                     L_U.inv, B.old, mu, W, weighted.delta.f) {

  C.t = t(C)

  i = 1

  for (i in 1:num.it) {
    # Allow for scaling of penalty on non-intercept variables (not applying to fusion)

    if (!is.null(penalty.factors)) {
      B.sparsity = B.old[1:p, ] * matrix(penalty.factors, p, k)
    } else {
      B.sparsity = B.old[1:p, ]
    }

    # Need to exclude the intercept variable from the B matrix to avoid penalization
    if (intercept) {
      B.sparsity = rbind(B.sparsity, 0)
      B.fusion = rbind(B.old[1:p, ], 0)
    } else {
      B.fusion = B.old
    }

    A.star = cbind(B.sparsity %*% C[, 1:k], B.fusion %*% C[, (k + 1):dim(C)[2]]) /
      mu

    # A.star = (B.star %*% C) / mu
    A.star[A.star > 1] = 1
    A.star[A.star < -1] = -1

    # Derivative is zero when taken with respect to each beta[,k.i]
    if (!is.null(XX)) {
      delta.lik <-
        lapply(1:k, function(k.i)
          (XX[[k.i]] %*% B.old[, k.i] - XY[[k.i]]) / samp.sizes[k.i])
    } else {
      delta.lik = lapply(1:k, function(k.i) {
        g = group.names[k.i]
        (crossprod(X.list[[k.i]], X.list[[k.i]] %*% B.old[, k.i]) - crossprod(X.list[[k.i]], Y[groups == g])) /
          samp.sizes[k.i]
      })
    }

    delta.f = do.call("cbind", delta.lik) + A.star %*% C.t

    B.new = W - L_U.inv * delta.f

    weighted.delta.f = weighted.delta.f + L_U.inv * 0.5 * i * delta.f

    Z = -weighted.delta.f

    W = (i * B.new + 2 * Z) / (i + 2)

    i = i + 1

    improvement = sum(abs(B.old - B.new))

    if (improvement < tol * p)
      break

    B.old = B.new
  }

  # Store the number of iterations
  .pkg.env$lastNumIters = i

  if (i >= num.it)
    warning("Reached max iterations without convergence.")

  colnames(B.new) = group.names

  return(B.new)
}







#' Fused lasso optimisation with proximal-gradient method.
#' (Chen et al. 2010)
#'
#' @param X matrix of covariates (n by p)
#' @param Y vector of responses (length n)
#' @param groups vector of group indicators (length n)
#' @param lambda Sparsity hyperparameter (accepts scalar value only)
#' @param gamma Fusion hyperparameter (accepts scalar value only)
#' @param G Matrix of pairwise group information sharing weights (K by K)
#' @param mu Smoothness parameter for proximal optimization
#' @param tol Tolerance for optimization
#' @param num.it Number of iterations
#' @param lam.max Maximal eigenvalue of \code{t(X) \%*\% X} (will be calculate
#' if not provided)
#' @param c.flag Whether to use Rcpp for certain calculations (see Details).
#' @param intercept Whether to include a (group-specific) intercept term.
#' @param penalty.factors Weights for sparsity penalty.
#' @param conserve.memory Whether to calculate XX and XY on the fly, conserving memory
#' at the cost of speed. (True by default iff p >= 10000)
#' @param scaling if TRUE, scale the sum-squared loss for each group by 1/n_k
#' where n_k is the number of samples in group k.
#'
#' @details
#' The proximal algorithm uses \code{t(X) \%*\% X} and \code{t(X) \%*\% Y}. The function will attempt to
#' pre-calculate these values to speed up computation. This may not always be possible due to
#' memory restrictions; at present this is only done for p < 10,000. When p > 10,000,
#' crossproducts are calculated explicitly; calculation can be speeded up by using
#' Rcpp code (setting c.flag=TRUE).
#'
#' @return
#' A matrix with the linear coefficients for each group (p by k).
#'
#' @export
#'
#' @examples
#' set.seed(123)
#' # Generate simple heterogeneous dataset
#' k = 4 # number of groups
#' p = 100 # number of covariates
#' n.group = 15 # number of samples per group
#' sigma = 0.05 # observation noise sd
#' groups = rep(1:k, each=n.group) # group indicators
#' # sparse linear coefficients
#' beta = matrix(0, p, k)
#' nonzero.ind = rbinom(p*k, 1, 0.025/k) # Independent coefficients
#' nonzero.shared = rbinom(p, 1, 0.025) # shared coefficients
#' beta[which(nonzero.ind==1)] = rnorm(sum(nonzero.ind), 1, 0.25)
#' beta[which(nonzero.shared==1),] = rnorm(sum(nonzero.shared), -1, 0.25)
#'
#' X = lapply(1:k,
#'            function(k.i) matrix(rnorm(n.group*p),
#'                                 n.group, p)) # covariates
#' y = sapply(1:k,
#'            function(k.i) X[[k.i]] %*% beta[,k.i] +
#'                            rnorm(n.group, 0, sigma)) # response
#' X = do.call('rbind', X)
#'
#' # Pairwise Fusion strength hyperparameters (tau(k,k'))
#' # Same for all pairs in this example
#' G = matrix(1, k, k)
#'
#' # Use L1 fusion to estimate betas (with near-optimal sparsity and
#' # information sharing among groups)
#' beta.estimate = fusedLassoProximal(X, y, groups, lambda=0.01, tol=3e-3,
#'                                    gamma=0.01, G, intercept=FALSE,
#'                                    num.it=500)
fusedLassoProximal <-
  function(X, Y, groups, lambda, gamma, G, mu = 1e-04, tol = 1e-06,
           num.it = 1000, lam.max = NULL, c.flag = FALSE, intercept = TRUE,
           penalty.factors = NULL, conserve.memory = p >= 10000, scaling=TRUE) {
    group.names = sort(unique(groups))

    # Only used if we are scaling the objective
    # function by 1/n_k for each group.
    samp.sizes = table(groups)[group.names]
    if(!scaling) samp.sizes[] = 1


    k = length(group.names)
    p = dim(X)[2]

    if (is.null(penalty.factors))
      penalty.factors = rep(1, dim(X)[2])

    if (intercept) {
      X = cbind(X, 1)
    }

    if (!conserve.memory) {
      # p < 10000 is a rough limit for keeping everything in memory
      XX.list = lapply(group.names, function(x) {
        indices = groups == x
        crossprod(X[indices, ])
      })

      XY.list = lapply(group.names, function(x) {
        indices = groups == x
        crossprod(X[indices, ], Y[indices])
      })
      X.list = NULL
    } else {
      # Otherwise calculate on the fly (takes longer, but saves memory)
      XX.list = NULL
      XY.list = NULL

      # Avoid indexing over large matrix
      X.list = list()

      for (k.i in 1:k) {
        X.list[[k.i]] = X[groups == group.names[k.i], ]
      }

      X = NULL # No longer need full matrix, delete to save memory
    }


    C.fusion = matrix(0, k, choose(k, 2))

    edge.i = 1

    for (i in 1:(k - 1)) {
      for (j in (i + 1):k) {
        C.fusion[i, edge.i] = G[i, j]
        C.fusion[j, edge.i] = -G[i, j]
        edge.i = edge.i + 1
      }
    }

    C.sparsity = lambda * diag(k)

    C = cbind(C.sparsity, gamma * C.fusion)

    return(
      .genFusedLassoProximal(
        X, Y, groups, group.names, XX.list, XY.list,
        p, k, samp.sizes, lambda, gamma, C, G, epsilon = mu * p * dim(C)[2],
        tol, num.it, lam.max, intercept = intercept, penalty.factors = penalty.factors,
        X.list = X.list, c.flag = c.flag
      )
    )
  }
